src/strapi-loader.ts

```diff
@@ -0,0 +1,72 @@
+import type { Loader, LoaderContext } from "astro/loaders";
+import { type API, strapi } from "@strapi/client";
+import invariant from "tiny-invariant";
+import type { StrapiLoaderOptions } from "./options";
+
+export function strapiLoader({
+  contentType,
+  clientConfig,
+  params,
+  pluralContentType = `${contentType}s`,
+  cacheDurationInMs = 0,
+  pageSize = 25,
+}: StrapiLoaderOptions): Loader {
+  const client = strapi(clientConfig);
+  const collection = client.collection(pluralContentType);
+  return {
+    name: contentType,
+    load: async function ({
+      store,
+      meta,
+      logger,
+      generateDigest,
+      parseData,
+    }: LoaderContext) {
+      const lastSynced = meta.get("lastSynced");
+
+      if (lastSynced && Date.now() - Number(lastSynced) < cacheDurationInMs) {
+        logger.info("Skipping sync");
+        return;
+      }
+
+      logger.debug(
+        `Fetching from Strapi with params: ${JSON.stringify(params)}`,
+      );
+
+      let currentPageNum = 1;
+      let totalPageCount = Number.MAX_SAFE_INTEGER; // TODO: is there a way to get this before paging?
+
+      do {
+        const paginatedParams = {
+          ...params,
+          pagination: { page: currentPageNum, pageSize },
+        } satisfies API.BaseQueryParams;
+        const {
+          data: page,
+          meta: { pagination },
+        } = await collection.find(paginatedParams);
+
+        for (const document of page) {
+          const id = String(document.id); // TODO: should this be documentId?
```

Comment:
I answered this question during my local testing. This should in fact be the `document.documentId`. If you make changes to the document, `document.id` increments, which in my book means it's _not_ an actual id. Unintuitive naming. But I need to change this.